lib/credentials/assume_role_with_web_identity_credentials.js

```diff
@@ -0,0 +1,153 @@
+var AWS = require('../core');
+var fs = require('fs');
+var crypto = require('crypto');
+var STS = require('../../clients/sts');
+
+/**
+ * Represents OIDC credentials from a file on disk
+ * If the credentials expire, the SDK can {refresh} the credentials
+ * from the file.
+ *
+ * ## Using the web identity token file
+ *
+ * This provider is checked by default in the Node.js environment. To use
+ * the provider simply add your OIDC token to a file (ASCII encoding) and
+ * share the filename in either AWS_WEB_IDENTITY_TOKEN_FILE environment
+ * variable or web_identity_token_file shared config variable
+ *
+ * The file contains encoded OIDC token and the characters are
+ * ASCII encoded. OIDC tokens are JSON Web Tokens (JWT).
+ * JWT's are 3 base64 encoded strings joined by the '.' character.
+ *
+ * This class will read filename from AWS_WEB_IDENTITY_TOKEN_FILE
+ * environment variable or web_identity_token_file shared config variable,
+ * and get the OIDC token from filename.
+ * It will also read IAM role to be assumed from AWS_IAM_ROLE_ARN
+ * environment variable or role_arn shared config variable.
+ * This provider gets credetials using the {AWS.STS.assumeRoleWithWebIdentity}
+ * service operation
+ *
+ * @!macro nobrowser
+ */
+AWS.AssumeRoleWithWebIdentityCredentials = AWS.util.inherit(AWS.Credentials, {
+
+  /**
+   * @overload AWS.AssumeRoleWithWebIdentityCredentials()
+   */
+  constructor: function AssumeRoleWithWebIdentityCredentials() {
```

Comment:
It's better to provide the construct parameters like other credential providers, like [here](https://github.com/aws/aws-sdk-js/blob/master/lib/credentials/web_identity_credentials.d.ts#L11). If we can set OIDC token, role arn, and session name from the constructor, it will be easier for us to debug the credential provider in the future by ignoring all underlying file loading and env vars. 
I can already see the feature request to set the config of `sts` client here. It will be handy if users can set STS config from the credentials provider constructor.

Reply:
As per the requirements, the parameters to assumeRoleWithWebIdentity should be read from either the environment variables or shared config.
WebIdentityCredentials already accepts parameters which should be passed to assumeRoleWithWebIdentity. Should I create two Credentials providers - one which accepts nothing and another one which accepts OIDC token file, role arn and session name?

Reply:
You don't have to. These constructor parameters should be all optional. And if they are set, they should be used without loading credentials from files and env vars. If they are undefined, we should follow the routine and load them as designed.

Reply:
Any naming standards to be followed for optional parameters as they're not defined in the requirements?
The parameters `RoleArn` and `RoleSessionName` could be the same as what's accepted by [assumeRoleWithWebIdentity](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/STS.html#assumeRoleWithWebIdentity-property). Assuming the third param would be named `WebIdentityTokenFile` which is similar to env variable `AWS_WEB_IDENTITY_TOKEN_FILE` or shared config key `web_identity_token_file`

Reply:
Providing an option to pass constructor parameters will affect typings.

For example:
* As per WebIdentityCredentials provider, the parameter RoleSessionName is optional https://github.com/aws/aws-sdk-js/blob/2627d1af01bf1a1dc745361c3edf9e825594dcd4/lib/credentials/web_identity_credentials.js#L44-L71
* However, in the type definitions when options are passed it should be similar to that of `STS.AssumeRoleWithWebIdentityRequest` https://github.com/aws/aws-sdk-js/blob/2627d1af01bf1a1dc745361c3edf9e825594dcd4/lib/credentials/web_identity_credentials.d.ts#L11
* And in `STS. AssumeRoleWithWebIdentityRequest`, the parameter `RoleSessionName` is not optional https://github.com/aws/aws-sdk-js/blob/57f295b969e7f195aae3ac3cd1a0fdb016093050/clients/sts.d.ts#L180-L192

I think we should introduce optional constructor parameters only when (and if) there's a requirement.

Reply:
As discussed offline, we won't be passing optional parameters in constructor, as it's not in the requirements

Reply:
As discussed offline, we decided to just add optional clientConfig as it's done in https://github.com/aws/aws-sdk-js/blob/2627d1af01bf1a1dc745361c3edf9e825594dcd4/lib/credentials/web_identity_credentials.js#L64

Reply:
Resolved in https://github.com/aws/aws-sdk-js/pull/2737/commits/e55f3a3114bdcfe5190f6a6102a54ad53b5b7254